### PR TITLE
backport: merge bitcoin#21523, #17954, #17905, #17993, #19132, #20494, #19425, #17513, bitcoin-gui#8, partial bitcoin#17938, #18587 (auxiliary backports: part 9)

### PR DESCRIPTION
**Pull request by kittywhiskers**
## Additional Notes

* Depends on https://github.com/dashpay/dash/pull/5235
* Depends on https://github.com/dashpay/dash/pull/5173

### CODE REVIEW

File: src/qt/walletmodel.cpp [hunk 1]
```diff
@@ -137,11 +136,6 @@ void WalletModel::updateChainLockHeight(int chainLockHeight)
     fForceCheckBalanceChanged = true;
 }
 
-int WalletModel::getNumBlocks() const
-{
-    return cachedNumBlocks;
-}
-
 int WalletModel::getNumISLocks() const
```
Comment (knst) [inline src/qt/walletmodel.cpp:139]: I think `cachedNumISLocks` and `getNumISLocks` should be nucked too.

File: test/lint/lint-circular-dependencies.sh [hunk 2]
```diff
@@ -56,6 +56,7 @@ EXPECTED_CIRCULAR_DEPENDENCIES=(
     "qt/bitcoingui -> qt/guiutil -> qt/bitcoingui"
     "qt/guiutil -> qt/optionsdialog -> qt/guiutil"
     "qt/guiutil -> qt/qvalidatedlineedit -> qt/guiutil"
+    "qt/clientmodel -> qt/guiutil -> qt/walletmodel -> qt/clientmodel"
```
Comment (knst) [inline test/lint/lint-circular-dependencies.sh:59]: just-a-note: this circular dependency due to missing bitcoin/bitcoin#17513

File: src/wallet/test/wallet_tests.cpp [hunk 3]
```diff
@@ -262,16 +252,11 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
...
-    node.mempool = std::make_unique<CTxMemPool>(node.fee_estimator.get());
-    auto chain = interfaces::MakeChain(node);
-
     std::string backup_file = (GetDataDir() / "wallet.backup").string();
 
     // Import key into wallet and call dumpwallet to create backup file.
     {
-        std::shared_ptr<CWallet> wallet = std::make_shared<CWallet>(chain.get(), "", CreateDummyWalletDatabase());
+        std::shared_ptr<CWallet> wallet = std::make_shared<CWallet>(m_node.chain.get(), "", CreateDummyWalletDatabase());
         auto spk_man = wallet->GetLegacyScriptPubKeyMan();
```
Comment (knst) [inline src/wallet/test/wallet_tests.cpp:260]: will conflict with #5173

**Review by knst — CHANGES_REQUESTED** [review]
1.
I think these 2 backports should be included in this PR:
 - Merge bitcoin#19132: qt: lock cs_main, m_cached_tip_mutex in that order
 - Merge bitcoin-core/gui#8: Fix regression in TransactionTableModel

Because they contains fixes for:
    Merge #17993: gui: Balance/TxStatus polling update based on last block hash.

2.
#19425 partial due to missing changes in these lines?
 
m_node.chain->broadcastTransaction 

(after comment "Load wallet again, this time creating new block and mempool")

Better to wait https://github.com/dashpay/dash/pull/5173 to be merged  and after that complete 19425.

3. 
I think better to avoid adding extra functions to interface of BaseHash in standard.h

**Comment by knst** [comment]
> I think better to avoid adding extra functions to interface of BaseHash in standard.h
sorry, forgot to attach a link.

Here is it:  [Nuke IsNull](https://github.com/knst/dash/commit/14c77b2b2fef54aa4e97b172790db8408012036c)

**Review by UdjinM6 — COMMENTED** [review]
LGTM, one question

File: src/init.cpp [hunk 4]
```diff
@@ -2194,7 +2194,7 @@ bool AppInitMain(const CoreContext& context, NodeContext& node, interfaces::Bloc
                         // work when we allow VerifyDB to be parameterized by chainstate.
                         if (&::ChainstateActive() == chainstate &&
                             !CVerifyDB().VerifyDB(
-                                chainparams, *chainstate, &chainstate->CoinsDB(),
+                                *chainstate, chainparams, chainstate->CoinsDB(),
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/init.cpp:2194]: Resolved with latest push, thanks for pointing it out!

File: test/lint/lint-circular-dependencies.sh (hunk 2 above)
Comment (kittywhiskers) [reply test/lint/lint-circular-dependencies.sh]: Resolved in latest push

File: src/wallet/test/wallet_tests.cpp (hunk 3 above)
Comment (kittywhiskers) [reply src/wallet/test/wallet_tests.cpp]: No longer conflicting as PR is merged

**Review by PastaPastaPasta — DISMISSED** [review]
light-ACK for merging via merge commit

**Comment by PastaPastaPasta** [comment]
(I guess we still need 1 more reviewer so sorry about the rebase, I was gonna merge it)

**Comment by PastaPastaPasta** [comment]
Not sure if this was introduced here but needs investigation
```
2023-04-06T05:43:33Z POTENTIAL DEADLOCK DETECTED
Previous lock order was:
 (2) 'm_cached_tip_mutex' in qt/clientmodel.cpp:150 (in thread 'main')
 (1) '::cs_main' in node/interfaces.cpp:325 (in thread 'main')
Current lock order is:
 'm_cs_chainstate' in validation.cpp:3194 (in thread 'msghand')
 (1) 'cs_main' in validation.cpp:3211 (in thread 'msghand')
 'm_mempool.cs' in validation.cpp:3212 (in thread 'msghand')
 (2) 'clientmodel->m_cached_tip_mutex' in qt/clientmodel.cpp:295 (in thread 'msghand')

Assertion failed: detected inconsistent lock order for 'clientmodel->m_cached_tip_mutex' in qt/clientmodel.cpp:295 (in thread 'msghand'), details in debug log.
2023-04-06T05:43:34Z Posix Signal: Abort trap: 6
   0#: (0x1051EB010) stacktraces.cpp         - HandlePosixSignal(int)
   1#: (0x1A5BFB2A4) <unknown-file>          - ???
   2#: (0x1A5BCCCEC) <unknown-file>          - ???
   3#: (0x1A5B062C8) <unknown-file>          - ???
   4#: (0x1051EFABC) sync.cpp:137            - EnterCritical(char const*, char const*, int, void*, bool)
   5#: (0x104BFF104) __mutex_base:209        - BlockTipChanged(ClientModel*, bool, interfaces::BlockTip, double, bool)
   6#: (0x104C00654) invoke.h                - ???
   7#: (0x104F7551C) function.h              - ???
   8#: (0x1051106D4) function.h              - ???
   9#: (0x105121964) optional.hpp:396        - ???
  10#: (0x10512160C) list:396                - ???
  11#: (0x10513F128) validation.cpp          - ???
  12#: (0x1051473B0) validation.cpp:4304     - ???
  13#: (0x104F373A8) net_processing.cpp:3773 - ???
  14#: (0x104F493FC) atomic:963              - ???
Posix Signal: Abort trap: 6
   0#: (0x1051EB010) stacktraces.cpp         - HandlePosixSignal(int)
   1#: (0x1A5BFB2A4) <unknown-file>          - ???
   2#: (0x1A5BCCCEC) <unknown-file>          - ???
   3#: (0x1A5B062C8) <unknown-file>          - ???
   4#: (0x1051EFABC) sync.cpp:137            - EnterCritical(char const*, char const*, int, void*, bool)
   5#: (0x104BFF104) __mutex_base:209        - BlockTipChanged(ClientModel*, bool, interfaces::BlockTip, double, bool)
   6#: (0x104C00654) invoke.h                - ???
   7#: (0x104F7551C) function.h              - ???
   8#: (0x1051106D4) function.h              - ???
   9#: (0x105121964) optional.hpp:396        - ???
  10#: (0x10512160C) list:396                - ???
  11#: (0x10513F128) validation.cpp          - ???
  12#: (0x1051473B0) validation.cpp:4304     - ???
  13#: (0x104F373A8) net_processing.cpp:3773 - ???
  14#: (0x104F493FC) atomic:963              - ???
[1]    97104 abort      ./src/qt/dash-qt --testnet --nowallet --wallet="mining_sql" --printtoconsole 
 ```

**Comment by UdjinM6** [comment]
@PastaPastaPasta good catch! 👍  backporting https://github.com/bitcoin/bitcoin/pull/19132 should fix it
and we might also need https://github.com/bitcoin-core/gui/pull/8 to fix another crash

**Comment by kittywhiskers** [comment]
@UdjinM6 @knst

**Review by knst — CHANGES_REQUESTED** [review]
Please, rename PR to something like:
backport: merge bitcoin#17513, #17905, #17938, #17954, #17993, partial #18587, #19132, #19425, #20494, #21523, bitcoin-gui#8 (auxiliary backports: part 9)

File: src/node/interfaces.cpp [hunk 5]
```diff
@@ -1,41 +1,473 @@
...
+#include <warnings.h>
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <coinjoin/coinjoin.h>
+#include <coinjoin/options.h>
+
+#include <univalue.h>
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/node/interfaces.cpp:57]: UniValue is an [external library](https://github.com/jgarzik/univalue) and even though in recent times, there seems to be an indication that despite a separate development history, it is very much centered around its use in Bitcoin Core, it is still a distinct entity.

Therefore their includes shouldn't be mixed with internal headers.

File: src/node/interfaces.cpp [hunk 6]
```diff
@@ -1,41 +1,473 @@
-// Copyright (c) 2018-2019 The Bitcoin Core developers
+// Copyright (c) 2018 The Bitcoin Core developers
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/node/interfaces.cpp]: The copyright year was inherited from `interfaces/node.cpp` but I've replaced with the copyright range from #20494 in the latest push

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -957,11 +957,12 @@ void CWallet::LoadToWallet(CWalletTx& wtxIn)
 {
     // If wallet doesn't have a chain (e.g dash-wallet), don't bother to update txn.
     if (HaveChain()) {
-        std::optional<int> block_height = chain().getBlockHeight(wtxIn.m_confirm.hashBlock);
-        if (block_height) {
+        bool active;
+        int height;
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/wallet/wallet.cpp:961]: Some Dash-specific cleanup is going to be done in a PR after this gets merged

**Review by knst — APPROVED** [review]
utACK

**Comment by UdjinM6** [comment]
why [284a20a](https://github.com/dashpay/dash/pull/5246/commits/284a20acbf7d171982928a547b9f57f7dd010e3f) is here? it looks unrelated.

**Comment by kittywhiskers** [comment]
Was running the extended test suite locally and found this test to be consistently failing, CI excludes `feature_pruning` and `feature_dbcrash` so at some point it broke but warnings weren't raised. Figured I'd add the fix into an "auxiliary" PR so it's here. It can be moved to a distinct PR if necessary but the fix was minor enough that I didn't really feel it should be.

**Comment by UdjinM6** [comment]
I'd say drop it cause it's not a backport and it's not a fix for any backport listed here

**Comment by kittywhiskers** [comment]
284a20a has been split off into https://github.com/dashpay/dash/pull/5324

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK

**Review by PastaPastaPasta — APPROVED** [review]
light-ACK for merging via merge commit

**Pull request merged**